### ISSUE TITLE
msi option runAfterFinish enabled: launched process gets wrong PATH environment variable

### ISSUE THREAD

**Issue by vincesp** [electron-userland/electron-builder#7979]
* **Electron-Builder Version**: 24.9.1

* **Node Version**: 18

* **Electron Version**: 28.0.0
* **Electron Type (current, beta, nightly)**: current

* **Target**: msi

I am creating an msi build with the default setting `runAfterFinish` enabled. However, when my process gets launched from the installer, it gets passed some kind of "Frankenstein" environment. While variables like `HOMEDRIVE`, `HOMEPATH`, `USERNAME`, `USERPROFILE`, or even `OneDrive` are inherited from the current user's environment, the `PATH` variable comes from the system, so it is missing crucial entries and my application crashes. The only fix is to disable `runAfterFinish`, unfortunately.

**Comment by vincesp** [comment]
Not stale

**Comment by beyondkmp** [comment]
@vincesp Can you provide a minimal demo to reproduce the issue?

**Comment by vincesp** [comment]
> @vincesp Can you provide a minimal demo to reproduce the issue?

@beyondkmp I created a simple example here, instructions in README: https://github.com/vincesp/electron-builder-issue7979

**Comment by beyondkmp** [comment]
This issue is caused by these codes.
https://github.com/electron-userland/electron-builder/blob/f4d40f91f1511fc55cbef7c9e7edfddaf6ab67bc/packages/app-builder-lib/templates/msi/template.xml#L29-L30

However, I haven't found a good solution yet. @mmaietta , do you have any suggestions?

**Comment by vincesp** [comment]
Still not stale, please reopen

**Comment by vincesp** [comment]
Not stale

**Comment by mmaietta** [comment]
Based on this @vincesp, I don't think it's possible to use the user's PATH env var since the app launch is in the scope of elevated/admin permissions (so admin user PATH would be used instead)
> During MSI installation, it elevated privileges and installed using the system user. Therefore, after startup, it uses the system user's environment variables and cannot access the current user's environment variables. Only when users actively click to open it can it access the current user's environment variables.

Happy to be proven wrong or discuss alternative suggestions, but closing this for now to clean up the backlog since there are no routes forward. If we're in the system user's context, how should it know which user's PATH env var to use? (especially if multiple accounts exist on the PC) AFAICT, this isn't possible to do so from electron-builder's side as any implementation requires custom work from the app side.

Here's the guidance via GPT

----------------

The short answer is: **no, not directly**. Once an **MSI installer is elevated**, it runs under a **different security context** (typically `NT AUTHORITY\SYSTEM`), and **cannot directly access the user's environment variables**, including their `PATH`.

### Details:

When an MSI installer elevates:

* It runs as **SYSTEM** or another administrative user.
* The `PATH` environment variable it sees is from the **system**, not the logged-in user's environment.

So if you try to read the `PATH` variable (e.g., from a `CustomAction`), it will return the system PATH, **not** the user's.

---

### Workarounds

1. **Use a deferred custom action with impersonation**:
   If you need to access the user's environment, you'd have to use a **non-elevated** (i.e., impersonated) custom action that runs in the **user context**. This is hard in MSI because:

   * Most custom actions are deferred and require elevation.
   * Impersonated custom actions have limitations (e.g., can’t modify system files).

2. **Schedule the action before elevation**:
   If you can read the user's `PATH` **before** elevation (i.e., during immediate actions), you can store it in a property and use it later.

3. **Install a small helper app** that runs post-install in the user context to modify their PATH, e.g., via:

   ```bat
   setx PATH "%PATH%;C:\Path\To\App"
   ```

   But note:

   * `setx` only modifies future sessions, not the current one.
   * This needs to run **as the user**, not from the MSI.

4. **Use a scheduled task or run-on-login script**:
   You can install a script or program that runs on user login and updates their environment appropriately.

---

### In summary:

> **An elevated MSI cannot access or modify the current user's PATH directly, because it is running as a different user (SYSTEM).**

You'd need to **separate user-context logic** from the elevated installer—e.g., via a helper run at login, or by pre-capturing the info in an unelevated context.
